package-builder/validate-build.bash

```diff
@@ -0,0 +1,409 @@
+        # If the path really doesn't exist in the build env...
+        if [[ ! -e $ref_path ]]; then
+            local -a hints=()
+            # Return nonzero.
+            rc=1
+
+            # compute the same relative suffix under develop_env
+            local rel="${ref_path#"$build_env"}"
+            local basename="${ref_path##*/}"
+            local dev_path="$develop_env$rel"
+
+            if [[ -e "$dev_path" ]]; then
+                # realpath gives the absolute, canonical path in the store
+                local real
+                real="$(realpath "$dev_path")"
+                # If we take the first 3 path elements we get the storepath
+                # of the package that provides the file.
+                local storepath
+                storepath="${real%/*/*}"
+                # Calculate hints by searching for realpath in closure of
+                # packages installed to develop environment.
+                local -A _hint_pkgs=()
+                find_packages_in_env _hint_pkgs "$storepath" "$develop_env"
+
+                # If we did find the file's package in any of the installed
+                # package closures then report those hints.
+                if [[ ${#_hint_pkgs[@]} -gt 0 ]]; then
+                    local -a hints=()
+                    for i in "${!_hint_pkgs[@]}"; do
+                        case "${_hint_pkgs[$i]}" in
+                            toplevel)
+                                # If the package is in the toplevel group, then
+                                # suggest adding it to runtime-packages.
+                                hints+=("consider adding package '$i' to 'runtime-packages'")
+                                ;;
+                            *)
+                                # Otherwise, suggest moving it to the toplevel group.
+                                hints+=("consider moving package '$i' to the 'toplevel' pkg-group")
+                                ;;
+                        esac
+                    done
+                else
+                    hints+=("report bug to Flox: '$dev_path' not referenced from manifest")
+                fi
+            else
+                hints+=("check your build script and project files for any mention of the '$basename' string")
+            fi
+            printf "❌ ERROR: Nonexistent path reference to '%s' found in package '%s':\n" \
+                "\$FLOX_ENV$rel" "$_pname"
+            for i in "${hints[@]}"; do
+                echo "    Hint: $i"
+            done
+            printf "%s\n%s\n\n" \
+                "Path referenced by: $src_file" \
+                "      Not found at: $ref_path"
+        fi
+    # Note: we assume filenames and ref_paths contain no colons.
+    done < <(
+        grep --binary-files=text -RHo -e "${build_env}[^:\"'[:cntrl:][:space:]]*" -- "$output_dir"
+    )
+
+    return $rc
+}
+
+#
+# main()
+#
+
+# Verify the requisites.
+if false; then verify_requisites >&2; fi
```

Comment:
Again, this `if false` is temporary pending the resolution of the problem #3311 was trying to fix.

Reply:
This temporary hack has been removed - resolving conversation.